## COMP541: DEEP LEARNING
## ASSIGNMENT 2: Convolutional Neural Network
### Fall 2022
### Gamze Keçibaş | 60211
---
**TABLE OF CONTENT**     
* ***Convolutional Neural Networks***  
  * Convolutional Filter Receptive Field    
  * Run the Pytorch ConvNet     
  * Add Pooling Layers     
  * Regularize the Network!  
    * Dropout   
    * Weight regularization 
    * Early stopping   
    * Learning rate scheduling 
  * Experiment with the Architecture 
  * Optimize the Architecture 
  * Test the final Architecture on Variations of Data   
* ***Transfer Learning*** 

***1) Convolutional Neural Networks***

In [1]:
import torch
import numpy
import math
from PIL import Image
import requests

In [6]:
ACTORS_TXT = "new_subset_actors.txt"
ACTRESS_TXT = "new_subset_actresses.txt"
actor_data = []
actress_data = []

with open(ACTORS_TXT, "r") as actors:
    for actor in actors:
        split_actor_info = actor.split()
        actor_data.append(split_actor_info)

with open(ACTRESS_TXT, "r") as actresses:
    for actress in actresses:
        split_actress_info = actress.split()
        actress_data.append(split_actress_info)

In [16]:
actor_images = {}
actress_images = {}
       
for actor in actor_data:
    actor_images[actor[-1]] = Image.open(requests.get(actor[4], stream=True).raw)
    
for actress in actress_data:
    actress_images[actress[-1]] = Image.open(requests.get(actress[4], stream=True).raw)

UnidentifiedImageError: cannot identify image file <_io.BytesIO object at 0x7ff7486498f0>

In [14]:
actress_images

{'962388de77e47001075e2b9d3bcea5067a0d31ef2e3d51e056033fb5f6a23bc0': 'http://static2.wikia.nocookie.net/__cb20100626051952/glee/images/4/4f/Dianna-Agron-dianna-agron-8731364-1937-2560.jpg',
 '1bbcabc38bdbb1f412f14a712dbdfb33bdcbd22330f4a5daecf4d71cbefc990b': 'http://www.hdwallpapersonly.com/wp-content/uploads/2013/06/Dianna-Agron-2.jpg',
 '377da13ce6b487c7c4cd6cd605ba8039effa4c52d59734122d7ecbd019c999cf': 'http://cdn04.cdn.justjared.com/wp-content/uploads/2011/08/agron-letterman/dianna-agron-letterman-02.jpg',
 '85a2ac66d74e450cc2ffb9bc84b4a38b0c724377d37b6577385d88999aff91eb': 'http://cdn01.cdn.justjared.com/wp-content/uploads/2012/09/agron-cite/dianna-agron-salma-hayek-la-cite-du-cinema-launch-02.jpg',
 '0126056b4df4f69504e26b056947998cac5ee5e6e5626ef3941b1725d8723974': 'http://images2.fanpop.com/images/photos/8300000/Dianna-Agron-dianna-agron-8334684-794-1200.jpg',
 '462a174fd8ff562e1d04960da4194ef379710c275dcdf834fe2a713624334a1f': 'http://images2.fanpop.com/image/photos/8800000/di

### PART - 1

In [ ]:
from CnnModel import SimpleCNN
from train import Trainer
from DataLoader import LoaderClass
import torch
from torch.nn import CrossEntropyLoss
from torch import optim
from torchvision import transforms, utils
import numpy as np
import os
from PIL import Image,ImageFile

LR = 1e-4
Momentum = 0.9 # If you use SGD with momentum
BATCH_SIZE = 16
USE_CUDA = True
POOLING = False
NUM_EPOCHS = 200
PATIENCE = -1
TRAIN_PERCENT = 0.8
VAL_PERCENT = 0.2
NUM_ARTISTS = 11
DATA_PATH = "./art_data/artists"
ImageFile.LOAD_TRUNCATED_IMAGES = True # Do not change this


def load_artist_data():
    data = []
    labels = []
    artists = [x for x in os.listdir(DATA_PATH) if x != '.DS_Store']
    print(artists)
    for folder in os.listdir(DATA_PATH):
        class_index = artists.index(folder)
        for image_name in os.listdir(DATA_PATH + "/" + folder):
            img = Image.open(DATA_PATH + "/" + folder + "/" + image_name)
            artist_label = (np.arange(NUM_ARTISTS) == class_index).astype(np.float32)
            data.append(np.array(img))
            labels.append(artist_label)
    shuffler = np.random.permutation(len(labels))
    data = np.array(data)[shuffler]
    labels = np.array(labels)[shuffler]

    length = len(data)
    val_size = int(length*0.2)
    val_data = data[0:val_size+1]
    train_data = data[val_size+1::]
    val_labels = labels[0:val_size+1]
    train_labels = labels[val_size+1::]
    print(val_labels)
    data_dict = {"train_data":train_data,"val_data":val_data}
    label_dict = {"train_labels":np.array(train_labels),"val_labels":np.array(val_labels)}

    return data_dict,label_dict

if __name__ == "__main__":
    data,labels = load_artist_data()
    model = SimpleCNN(use_cuda=True,pooling=POOLING)
    optimizer = optim.Adam(model.parameters(), lr=LR)
    transforms = {
        'train': transforms.Compose([
            transforms.Resize(50),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        ]),
        'val': transforms.Compose([
            transforms.Resize(50),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        ]),
        }
    train_dataset = LoaderClass(data,labels,"train",transforms["train"])
    valid_dataset = LoaderClass(data,labels,"val",transforms["val"])

    train_loader = torch.utils.data.DataLoader(train_dataset,
                                               batch_size=BATCH_SIZE,
                                               shuffle=True, num_workers=4, pin_memory=True)
    val_loader = torch.utils.data.DataLoader(valid_dataset,
                                             batch_size=BATCH_SIZE,
                                             shuffle=True, num_workers=4, pin_memory=True)

    criterion = CrossEntropyLoss()
    trainer_m = Trainer(model, criterion, train_loader, val_loader, optimizer, num_epoch=NUM_EPOCHS, patience=PATIENCE,batch_size=BATCH_SIZE,lr_scheduler= None)
    best_model = trainer_m.train()

['richard wilson', 'paul cezanne', 'paul sandby', 'claude monet', 'peter paul rubens', 'canaletto', 'j. m. w. turner', 'paul gauguin', 'george romney', 'john robert cozens', 'rembrandt']
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 1. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]]


/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:566: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))
Epoch 0,phase: Training: 1it [00:00, 36472.21it/s, train_loss : ?, train_acc: ?]/content/drive/MyDrive/COLABS/02-convolutional_neural_network/comp541-hw2-starter-code/train.py:65: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  softmaxed_scores = self.softmax(logits)
Epoch 1,phase: Training: 2it [00:07,  3.91s/it, val acc:  0.323,val loss:  2.367]  

New best loss, loss is:  2.3667938709259033 acc is:  0.34065934065934067


Epoch 2,phase: Training: 3it [00:08,  2.64s/it, val acc:  0.323,val loss:  2.246]  

New best loss, loss is:  2.246028741200765 acc is:  0.34065934065934067


Epoch 3,phase: Training: 4it [00:09,  1.98s/it, val acc:  0.323,val loss:  2.186]  

New best loss, loss is:  2.186363101005554 acc is:  0.34065934065934067


Epoch 4,phase: Training: 5it [00:10,  1.59s/it, val acc:  0.344,val loss:  2.175]  

New best loss, loss is:  2.175256848335266 acc is:  0.3626373626373626


Epoch 5,phase: Training: 6it [00:11,  1.35s/it, val acc:  0.365,val loss:  2.163]  

New best loss, loss is:  2.1633835633595786 acc is:  0.38461538461538464


Epoch 6,phase: Training: 7it [00:12,  1.20s/it, val acc:  0.365,val loss:  2.154]  

New best loss, loss is:  2.15433136622111 acc is:  0.38461538461538464


Epoch 7,phase: Training: 8it [00:13,  1.09s/it, val acc:  0.365,val loss:  2.148]  

New best loss, loss is:  2.1482286850611367 acc is:  0.38461538461538464


Epoch 8,phase: Training: 9it [00:13,  1.02s/it, val acc:  0.375,val loss:  2.130]  

New best loss, loss is:  2.1295151909192405 acc is:  0.3956043956043956


Epoch 9,phase: Training: 10it [00:14,  1.02it/s, val acc:  0.417,val loss:  2.111]  

New best loss, loss is:  2.1109034220377603 acc is:  0.43956043956043955


Epoch 10,phase: Training: 11it [00:15,  1.05it/s, val acc:  0.448,val loss:  2.114] 

Not a better score


Epoch 11,phase: Training: 12it [00:16,  1.08it/s, val acc:  0.448,val loss:  2.097]  

New best loss, loss is:  2.0969110131263733 acc is:  0.4725274725274725


Epoch 12,phase: Training: 13it [00:17,  1.09it/s, val acc:  0.448,val loss:  2.099]  

Not a better score


Epoch 13,phase: Training: 14it [00:18,  1.11it/s, val acc:  0.427,val loss:  2.104]  

Not a better score


Epoch 14,phase: Training: 15it [00:19,  1.11it/s, val acc:  0.438,val loss:  2.089]  

New best loss, loss is:  2.0892046689987183 acc is:  0.46153846153846156


Epoch 15,phase: Training: 16it [00:20,  1.12it/s, val acc:  0.438,val loss:  2.086]  

New best loss, loss is:  2.085578123728434 acc is:  0.46153846153846156


Epoch 16,phase: Training: 17it [00:20,  1.13it/s, val acc:  0.458,val loss:  2.066]  

New best loss, loss is:  2.066051721572876 acc is:  0.4835164835164835


Epoch 17,phase: Training: 18it [00:21,  1.13it/s, val acc:  0.438,val loss:  2.085]  

Not a better score


Epoch 18,phase: Training: 19it [00:22,  1.13it/s, val acc:  0.458,val loss:  2.064]  

New best loss, loss is:  2.0643369356791177 acc is:  0.4835164835164835


Epoch 19,phase: Training: 20it [00:23,  1.13it/s, val acc:  0.469,val loss:  2.068]  

Not a better score


Epoch 20,phase: Training: 21it [00:24,  1.12it/s, val acc:  0.469,val loss:  2.052]  

New best loss, loss is:  2.051508148511251 acc is:  0.4945054945054945


Epoch 21,phase: Training: 22it [00:25,  1.13it/s, val acc:  0.500,val loss:  2.047]  

New best loss, loss is:  2.047407786051432 acc is:  0.5274725274725275


Epoch 22,phase: Training: 23it [00:26,  1.12it/s, val acc:  0.490,val loss:  2.044]  

New best loss, loss is:  2.0435296297073364 acc is:  0.5164835164835165


Epoch 23,phase: Training: 24it [00:27,  1.12it/s, val acc:  0.479,val loss:  2.051]  

Not a better score


Epoch 24,phase: Training: 25it [00:28,  1.12it/s, val acc:  0.490,val loss:  2.043]  

New best loss, loss is:  2.042909860610962 acc is:  0.5164835164835165


Epoch 25,phase: Training: 26it [00:28,  1.12it/s, val acc:  0.469,val loss:  2.037]  

New best loss, loss is:  2.037234365940094 acc is:  0.4945054945054945


Epoch 26,phase: Training: 27it [00:29,  1.12it/s, val acc:  0.479,val loss:  2.041]  

Not a better score


Epoch 27,phase: Training: 28it [00:30,  1.13it/s, val acc:  0.448,val loss:  2.041]  

Not a better score


Epoch 28,phase: Training: 29it [00:31,  1.13it/s, val acc:  0.458,val loss:  2.034]  

New best loss, loss is:  2.034455160299937 acc is:  0.4835164835164835


Epoch 29,phase: Training: 30it [00:32,  1.13it/s, val acc:  0.479,val loss:  2.040]  

Not a better score


Epoch 30,phase: Training: 31it [00:33,  1.14it/s, val acc:  0.469,val loss:  2.035]  

Not a better score


Epoch 31,phase: Training: 32it [00:34,  1.03it/s, val acc:  0.479,val loss:  2.042]  

Not a better score


Epoch 32,phase: Training: 33it [00:35,  1.11s/it, val acc:  0.479,val loss:  2.038]  

Not a better score


Epoch 33,phase: Training: 34it [00:37,  1.16s/it, val acc:  0.479,val loss:  2.048]  

Not a better score


Epoch 34,phase: Training: 35it [00:38,  1.20s/it, val acc:  0.490,val loss:  2.034]  

New best loss, loss is:  2.034157911936442 acc is:  0.5164835164835165


Epoch 35,phase: Training: 36it [00:39,  1.13s/it, val acc:  0.479,val loss:  2.050]  

Not a better score


Epoch 36,phase: Training: 37it [00:40,  1.09s/it, val acc:  0.479,val loss:  2.039]  

Not a better score


Epoch 37,phase: Training: 38it [00:41,  1.05s/it, val acc:  0.469,val loss:  2.045]  

Not a better score


Epoch 38,phase: Training: 39it [00:42,  1.01s/it, val acc:  0.479,val loss:  2.036]  

Not a better score


Epoch 39,phase: Training: 40it [00:43,  1.03it/s, val acc:  0.469,val loss:  2.062]  

Not a better score


Epoch 40,phase: Training: 41it [00:44,  1.06it/s, val acc:  0.479,val loss:  2.037]  

Not a better score


Epoch 41,phase: Training: 42it [00:45,  1.07it/s, val acc:  0.479,val loss:  2.067]  

Not a better score


Epoch 42,phase: Training: 43it [00:45,  1.08it/s, val acc:  0.469,val loss:  2.048]  

Not a better score


Epoch 43,phase: Training: 44it [00:46,  1.08it/s, val acc:  0.469,val loss:  2.046]  

Not a better score


Epoch 44,phase: Training: 45it [00:47,  1.08it/s, val acc:  0.479,val loss:  2.044]  

Not a better score


Epoch 45,phase: Training: 46it [00:48,  1.09it/s, val acc:  0.469,val loss:  2.048]  

Not a better score


Epoch 46,phase: Training: 47it [00:49,  1.09it/s, val acc:  0.448,val loss:  2.058]  

Not a better score


Epoch 47,phase: Training: 48it [00:50,  1.10it/s, val acc:  0.490,val loss:  2.048]  

Not a better score


Epoch 48,phase: Training: 49it [00:51,  1.10it/s, val acc:  0.490,val loss:  2.051]  

Not a better score


Epoch 49,phase: Training: 50it [00:52,  1.10it/s, val acc:  0.448,val loss:  2.060]  

Not a better score


Epoch 50,phase: Training: 51it [00:53,  1.08it/s, val acc:  0.458,val loss:  2.056]  

Not a better score


Epoch 51,phase: Training: 52it [00:54,  1.08it/s, val acc:  0.458,val loss:  2.050]  

Not a better score


Epoch 52,phase: Training: 53it [00:55,  1.08it/s, val acc:  0.427,val loss:  2.068]  

Not a better score


Epoch 53,phase: Training: 54it [00:56,  1.08it/s, val acc:  0.469,val loss:  2.062]  

Not a better score


Epoch 54,phase: Training: 55it [00:57,  1.07it/s, val acc:  0.438,val loss:  2.064]  

Not a better score


Epoch 55,phase: Training: 56it [00:57,  1.07it/s, val acc:  0.438,val loss:  2.049]  

Not a better score


Epoch 56,phase: Training: 57it [00:58,  1.07it/s, val acc:  0.448,val loss:  2.068]  

Not a better score


Epoch 57,phase: Training: 58it [00:59,  1.07it/s, val acc:  0.438,val loss:  2.062]  

Not a better score


Epoch 58,phase: Training: 59it [01:00,  1.08it/s, val acc:  0.448,val loss:  2.049]  

Not a better score


Epoch 59,phase: Training: 60it [01:01,  1.07it/s, val acc:  0.469,val loss:  2.060]  

Not a better score


Epoch 60,phase: Training: 61it [01:02,  1.08it/s, val acc:  0.448,val loss:  2.068]  

Not a better score


Epoch 61,phase: Training: 62it [01:03,  1.09it/s, val acc:  0.448,val loss:  2.069]  

Not a better score


Epoch 62,phase: Training: 63it [01:04,  1.09it/s, val acc:  0.438,val loss:  2.059]  

Not a better score


Epoch 63,phase: Training: 64it [01:05,  1.09it/s, val acc:  0.448,val loss:  2.065]  

Not a better score


Epoch 64,phase: Training: 65it [01:06,  1.10it/s, val acc:  0.448,val loss:  2.060]  

Not a better score


Epoch 65,phase: Training: 66it [01:07,  1.09it/s, val acc:  0.448,val loss:  2.058]  

Not a better score


Epoch 66,phase: Training: 67it [01:08,  1.10it/s, val acc:  0.417,val loss:  2.069]  

Not a better score


Epoch 67,phase: Training: 68it [01:08,  1.10it/s, val acc:  0.438,val loss:  2.065]  

Not a better score


Epoch 68,phase: Training: 69it [01:09,  1.09it/s, val acc:  0.448,val loss:  2.061]  

Not a better score


Epoch 69,phase: Training: 70it [01:10,  1.09it/s, val acc:  0.448,val loss:  2.064]  

Not a better score


Epoch 70,phase: Training: 71it [01:11,  1.10it/s, val acc:  0.438,val loss:  2.060]  

Not a better score


Epoch 71,phase: Training: 72it [01:12,  1.11it/s, val acc:  0.427,val loss:  2.072]  

Not a better score


Epoch 72,phase: Training: 73it [01:13,  1.11it/s, val acc:  0.438,val loss:  2.059]  

Not a better score


Epoch 73,phase: Training: 74it [01:14,  1.11it/s, val acc:  0.438,val loss:  2.067]  

Not a better score


Epoch 74,phase: Training: 75it [01:15,  1.11it/s, val acc:  0.438,val loss:  2.068]  

Not a better score


Epoch 75,phase: Training: 76it [01:16,  1.11it/s, val acc:  0.469,val loss:  2.075]  

Not a better score


Epoch 76,phase: Training: 77it [01:17,  1.11it/s, val acc:  0.448,val loss:  2.074]  

Not a better score


Epoch 77,phase: Training: 78it [01:17,  1.11it/s, val acc:  0.448,val loss:  2.072]  

Not a better score


Epoch 78,phase: Training: 79it [01:18,  1.11it/s, val acc:  0.448,val loss:  2.067]  

Not a better score


Epoch 79,phase: Training: 80it [01:19,  1.11it/s, val acc:  0.448,val loss:  2.060]  

Not a better score


Epoch 80,phase: Training: 81it [01:20,  1.11it/s, val acc:  0.438,val loss:  2.070]  

Not a better score


Epoch 81,phase: Training: 82it [01:21,  1.11it/s, val acc:  0.438,val loss:  2.057]  

Not a better score


Epoch 82,phase: Training: 83it [01:22,  1.09it/s, val acc:  0.438,val loss:  2.058]  

Not a better score


Epoch 83,phase: Training: 84it [01:23,  1.10it/s, val acc:  0.448,val loss:  2.062]  

Not a better score


Epoch 84,phase: Training: 85it [01:24,  1.10it/s, val acc:  0.448,val loss:  2.073]  

Not a better score


Epoch 85,phase: Training: 86it [01:25,  1.04it/s, val acc:  0.448,val loss:  2.063]  

Not a better score


Epoch 86,phase: Training: 87it [01:26,  1.09s/it, val acc:  0.438,val loss:  2.080]  

Not a better score


Epoch 87,phase: Training: 88it [01:28,  1.18s/it, val acc:  0.448,val loss:  2.079]  

Not a better score


Epoch 88,phase: Training: 89it [01:29,  1.17s/it, val acc:  0.448,val loss:  2.059]  

Not a better score


Epoch 89,phase: Training: 90it [01:30,  1.10s/it, val acc:  0.438,val loss:  2.095]  

Not a better score


Epoch 90,phase: Training: 91it [01:31,  1.05s/it, val acc:  0.469,val loss:  2.051]  

Not a better score


Epoch 91,phase: Training: 92it [01:32,  1.02s/it, val acc:  0.427,val loss:  2.070]  

Not a better score


Epoch 92,phase: Training: 93it [01:33,  1.01it/s, val acc:  0.438,val loss:  2.074]  

Not a better score


Epoch 93,phase: Training: 94it [01:34,  1.03it/s, val acc:  0.438,val loss:  2.079]  

Not a better score


Epoch 94,phase: Training: 95it [01:34,  1.04it/s, val acc:  0.448,val loss:  2.068]  

Not a better score


Epoch 95,phase: Training: 96it [01:35,  1.04it/s, val acc:  0.438,val loss:  2.059]  

Not a better score


Epoch 96,phase: Training: 97it [01:36,  1.04it/s, val acc:  0.438,val loss:  2.070]  

Not a better score


Epoch 97,phase: Training: 98it [01:37,  1.04it/s, val acc:  0.448,val loss:  2.074]  

Not a better score


Epoch 98,phase: Training: 99it [01:38,  1.05it/s, val acc:  0.438,val loss:  2.063]  

Not a better score


Epoch 99,phase: Training: 100it [01:39,  1.05it/s, val acc:  0.448,val loss:  2.055]  

Not a better score


Epoch 100,phase: Training: 101it [01:40,  1.05it/s, val acc:  0.448,val loss:  2.072] 

Not a better score


Epoch 101,phase: Training: 102it [01:41,  1.05it/s, val acc:  0.448,val loss:  2.063]  

Not a better score


Epoch 102,phase: Training: 103it [01:42,  1.05it/s, val acc:  0.438,val loss:  2.064]  

Not a better score


Epoch 103,phase: Training: 104it [01:43,  1.06it/s, val acc:  0.448,val loss:  2.073]  

Not a better score


Epoch 104,phase: Training: 105it [01:44,  1.06it/s, val acc:  0.448,val loss:  2.075]  

Not a better score


Epoch 105,phase: Training: 106it [01:45,  1.06it/s, val acc:  0.448,val loss:  2.061]  

Not a better score


Epoch 106,phase: Training: 107it [01:46,  1.06it/s, val acc:  0.448,val loss:  2.041]  

Not a better score


Epoch 107,phase: Training: 108it [01:47,  1.07it/s, val acc:  0.448,val loss:  2.048]  

Not a better score


Epoch 108,phase: Training: 109it [01:48,  1.07it/s, val acc:  0.427,val loss:  2.070]  

Not a better score


Epoch 109,phase: Training: 110it [01:49,  1.07it/s, val acc:  0.448,val loss:  2.069]  

Not a better score


Epoch 110,phase: Training: 111it [01:50,  1.08it/s, val acc:  0.448,val loss:  2.071]  

Not a better score


Epoch 111,phase: Training: 112it [01:51,  1.07it/s, val acc:  0.448,val loss:  2.074]  

Not a better score


Epoch 112,phase: Training: 113it [01:51,  1.06it/s, val acc:  0.438,val loss:  2.073]  

Not a better score


Epoch 113,phase: Training: 114it [01:52,  1.06it/s, val acc:  0.438,val loss:  2.062]  

Not a better score


Epoch 114,phase: Training: 115it [01:53,  1.06it/s, val acc:  0.448,val loss:  2.059]  

Not a better score


Epoch 115,phase: Training: 116it [01:54,  1.05it/s, val acc:  0.448,val loss:  2.067]  

Not a better score


Epoch 116,phase: Training: 117it [01:55,  1.05it/s, val acc:  0.448,val loss:  2.063]  

Not a better score


Epoch 117,phase: Training: 118it [01:56,  1.05it/s, val acc:  0.427,val loss:  2.080]  

Not a better score


Epoch 118,phase: Training: 119it [01:57,  1.04it/s, val acc:  0.448,val loss:  2.074]  

Not a better score


Epoch 119,phase: Training: 120it [01:58,  1.04it/s, val acc:  0.448,val loss:  2.062]  

Not a better score


Epoch 120,phase: Training: 121it [01:59,  1.04it/s, val acc:  0.438,val loss:  2.083]  

Not a better score


Epoch 121,phase: Training: 122it [02:00,  1.05it/s, val acc:  0.448,val loss:  2.081]  

Not a better score


Epoch 122,phase: Training: 123it [02:01,  1.05it/s, val acc:  0.438,val loss:  2.070]  

Not a better score


Epoch 123,phase: Training: 124it [02:02,  1.05it/s, val acc:  0.448,val loss:  2.064]  

Not a better score


Epoch 124,phase: Training: 125it [02:03,  1.06it/s, val acc:  0.438,val loss:  2.071]  

Not a better score


Epoch 125,phase: Training: 126it [02:04,  1.06it/s, val acc:  0.438,val loss:  2.080]  

Not a better score


Epoch 126,phase: Training: 127it [02:05,  1.06it/s, val acc:  0.438,val loss:  2.067]  

Not a better score


Epoch 127,phase: Training: 128it [02:06,  1.05it/s, val acc:  0.448,val loss:  2.079]  

Not a better score


Epoch 128,phase: Training: 129it [02:07,  1.05it/s, val acc:  0.448,val loss:  2.083]  

Not a better score


Epoch 129,phase: Training: 130it [02:08,  1.06it/s, val acc:  0.438,val loss:  2.074]  

Not a better score


Epoch 130,phase: Training: 131it [02:09,  1.08s/it, val acc:  0.438,val loss:  2.083]  

Not a better score


Epoch 131,phase: Training: 132it [02:11,  1.20s/it, val acc:  0.448,val loss:  2.069]  

Not a better score


Epoch 132,phase: Training: 133it [02:12,  1.25s/it, val acc:  0.438,val loss:  2.078]  

Not a better score


Epoch 133,phase: Training: 134it [02:13,  1.19s/it, val acc:  0.438,val loss:  2.069]  

Not a better score


Epoch 134,phase: Training: 135it [02:14,  1.26s/it, val acc:  0.448,val loss:  2.088]  

Not a better score


Epoch 135,phase: Training: 136it [02:16,  1.29s/it, val acc:  0.448,val loss:  2.075]  

Not a better score


Epoch 136,phase: Training: 137it [02:17,  1.33s/it, val acc:  0.438,val loss:  2.071]  

Not a better score


Epoch 137,phase: Training: 138it [02:18,  1.22s/it, val acc:  0.448,val loss:  2.089]  

Not a better score


Epoch 138,phase: Training: 139it [02:19,  1.14s/it, val acc:  0.438,val loss:  2.085]  

Not a better score


Epoch 139,phase: Training: 140it [02:20,  1.09s/it, val acc:  0.438,val loss:  2.081]  

Not a better score


Epoch 140,phase: Training: 141it [02:21,  1.05s/it, val acc:  0.438,val loss:  2.084]  

Not a better score


Epoch 141,phase: Training: 142it [02:22,  1.03s/it, val acc:  0.438,val loss:  2.078]  

Not a better score


Epoch 142,phase: Training: 143it [02:23,  1.00s/it, val acc:  0.438,val loss:  2.061]  

Not a better score


Epoch 143,phase: Training: 144it [02:24,  1.02it/s, val acc:  0.427,val loss:  2.077]  

Not a better score


Epoch 144,phase: Training: 145it [02:25,  1.03it/s, val acc:  0.448,val loss:  2.065]  

Not a better score


Epoch 145,phase: Training: 146it [02:26,  1.04it/s, val acc:  0.448,val loss:  2.079]  

Not a better score


Epoch 146,phase: Training: 147it [02:27,  1.04it/s, val acc:  0.427,val loss:  2.075]  

Not a better score


Epoch 147,phase: Training: 148it [02:28,  1.04it/s, val acc:  0.448,val loss:  2.089]  

Not a better score


Epoch 148,phase: Training: 149it [02:29,  1.04it/s, val acc:  0.438,val loss:  2.084]  

Not a better score


Epoch 149,phase: Training: 150it [02:30,  1.05it/s, val acc:  0.406,val loss:  2.092]  

Not a better score


Epoch 150,phase: Training: 151it [02:31,  1.03it/s, val acc:  0.438,val loss:  2.070]  

Not a better score


Epoch 151,phase: Training: 152it [02:31,  1.04it/s, val acc:  0.438,val loss:  2.066]  

Not a better score


Epoch 152,phase: Training: 153it [02:32,  1.04it/s, val acc:  0.469,val loss:  2.055]  

Not a better score


Epoch 153,phase: Training: 154it [02:33,  1.05it/s, val acc:  0.458,val loss:  2.066]  

Not a better score


Epoch 154,phase: Training: 155it [02:34,  1.05it/s, val acc:  0.448,val loss:  2.073]  

Not a better score


Epoch 155,phase: Training: 156it [02:35,  1.04it/s, val acc:  0.448,val loss:  2.056]  

Not a better score


Epoch 156,phase: Training: 157it [02:36,  1.05it/s, val acc:  0.406,val loss:  2.110]  

Not a better score


Epoch 157,phase: Training: 158it [02:37,  1.04it/s, val acc:  0.427,val loss:  2.096]  

Not a better score


Epoch 158,phase: Training: 159it [02:38,  1.05it/s, val acc:  0.458,val loss:  2.063]  

Not a better score


Epoch 159,phase: Training: 160it [02:39,  1.05it/s, val acc:  0.438,val loss:  2.070]  

Not a better score


Epoch 160,phase: Training: 161it [02:40,  1.04it/s, val acc:  0.438,val loss:  2.071]  

Not a better score


Epoch 161,phase: Training: 162it [02:41,  1.05it/s, val acc:  0.438,val loss:  2.070]  

Not a better score


Epoch 162,phase: Training: 163it [02:42,  1.05it/s, val acc:  0.438,val loss:  2.079]  

Not a better score


Epoch 163,phase: Training: 164it [02:43,  1.05it/s, val acc:  0.448,val loss:  2.052]  

Not a better score


Epoch 164,phase: Training: 165it [02:44,  1.05it/s, val acc:  0.458,val loss:  2.054]  

Not a better score


Epoch 165,phase: Training: 166it [02:45,  1.05it/s, val acc:  0.438,val loss:  2.058]  

Not a better score


Epoch 166,phase: Training: 167it [02:46,  1.04it/s, val acc:  0.438,val loss:  2.068]  

Not a better score


Epoch 167,phase: Training: 168it [02:47,  1.04it/s, val acc:  0.438,val loss:  2.084]  

Not a better score


Epoch 168,phase: Training: 169it [02:48,  1.04it/s, val acc:  0.458,val loss:  2.077]  

Not a better score


Epoch 169,phase: Training: 170it [02:49,  1.04it/s, val acc:  0.438,val loss:  2.088]  

Not a better score


Epoch 170,phase: Training: 171it [02:50,  1.04it/s, val acc:  0.438,val loss:  2.068]  

Not a better score


Epoch 171,phase: Training: 172it [02:51,  1.04it/s, val acc:  0.448,val loss:  2.082]  

Not a better score


Epoch 172,phase: Training: 173it [02:52,  1.03it/s, val acc:  0.427,val loss:  2.080]  

Not a better score


Epoch 173,phase: Training: 174it [02:53,  1.02it/s, val acc:  0.438,val loss:  2.079]  

Not a better score


Epoch 174,phase: Training: 175it [02:54,  1.03it/s, val acc:  0.448,val loss:  2.079]  

Not a better score


Epoch 175,phase: Training: 176it [02:55,  1.03it/s, val acc:  0.448,val loss:  2.069]  

Not a better score


Epoch 176,phase: Training: 177it [02:56,  1.04it/s, val acc:  0.427,val loss:  2.058]  

Not a better score


Epoch 177,phase: Training: 178it [02:56,  1.04it/s, val acc:  0.448,val loss:  2.061]  

Not a better score


Epoch 178,phase: Training: 179it [02:57,  1.04it/s, val acc:  0.448,val loss:  2.072]  

Not a better score


Epoch 179,phase: Training: 180it [02:58,  1.02it/s, val acc:  0.417,val loss:  2.077]  

Not a better score


Epoch 180,phase: Training: 181it [02:59,  1.02it/s, val acc:  0.448,val loss:  2.065]  

Not a better score


Epoch 181,phase: Training: 182it [03:00,  1.02it/s, val acc:  0.448,val loss:  2.065]  

Not a better score


Epoch 182,phase: Training: 183it [03:01,  1.03it/s, val acc:  0.448,val loss:  2.070]  

Not a better score


Epoch 183,phase: Training: 184it [03:03,  1.09s/it, val acc:  0.417,val loss:  2.086]  

Not a better score


Epoch 184,phase: Training: 185it [03:04,  1.21s/it, val acc:  0.448,val loss:  2.065]  

Not a better score


Epoch 185,phase: Training: 186it [03:06,  1.27s/it, val acc:  0.427,val loss:  2.063]  

Not a better score


Epoch 186,phase: Training: 187it [03:07,  1.20s/it, val acc:  0.469,val loss:  2.058]  

Not a better score


Epoch 187,phase: Training: 188it [03:08,  1.12s/it, val acc:  0.458,val loss:  2.044]  

Not a better score


Epoch 188,phase: Training: 189it [03:09,  1.07s/it, val acc:  0.427,val loss:  2.065]  

Not a better score


Epoch 189,phase: Training: 190it [03:10,  1.04s/it, val acc:  0.448,val loss:  2.050]  

Not a better score


Epoch 190,phase: Training: 191it [03:11,  1.03s/it, val acc:  0.479,val loss:  2.054]  

Not a better score


Epoch 191,phase: Training: 192it [03:11,  1.00s/it, val acc:  0.448,val loss:  2.063]  

Not a better score


Epoch 192,phase: Training: 193it [03:12,  1.01it/s, val acc:  0.448,val loss:  2.068]  

Not a better score


Epoch 193,phase: Training: 194it [03:13,  1.02it/s, val acc:  0.458,val loss:  2.056]  

Not a better score


Epoch 194,phase: Training: 195it [03:14,  1.03it/s, val acc:  0.469,val loss:  2.045]  

Not a better score


Epoch 195,phase: Training: 196it [03:15,  1.03it/s, val acc:  0.458,val loss:  2.056]  

Not a better score


Epoch 196,phase: Training: 197it [03:16,  1.03it/s, val acc:  0.448,val loss:  2.048]  

Not a better score


Epoch 197,phase: Training: 198it [03:17,  1.03it/s, val acc:  0.458,val loss:  2.064]  

Not a better score


Epoch 198,phase: Training: 199it [03:18,  1.03it/s, val acc:  0.458,val loss:  2.059]  

Not a better score


Epoch 199,phase: Training: 200it [03:19,  1.03it/s, val acc:  0.448,val loss:  2.062]  

Not a better score


Epoch 199,phase: Validation: 200it [03:20,  1.00s/it, val acc:  0.448,val loss:  2.058]

Not a better score
Training ended returning the best model
Best val acc: 0.5164835164835165, Best val loss: 2.034157911936442, Best train acc: 0.65, Best train loss: 1.894924241563548 


***1.1) Convolutional Filter Receptive Field***
* **What are the dimensions of the receptive field for a node in Z2? Note that you can ignore padding, and just consider patches in the middle of the image and Z1.** 



* **Why is it effective to build convolutional networks deeper, i.e. with more layers?**

***1.2) Run the Pytorch ConvNet***      
**• How many layers are there? Are they all convolutional? If not, what structure do they have?** 
The best model from run.py has six layers. First one is 2 dimensional convolutional layer, second layer is 2 dimensional max pooling layer, third one is 2 dimensional convolutional, forth one is 2 dimensional max pooling and final two layers are fully connected linear layers.    
  
**• Which activation function is used on the hidden nodes?**  
Rectified Linear Unit (ReLU) is used to on the hidden nodes.  
  
**• What loss function is being used to train the network?**  
Cross Entropy Loss (softmax) is preferred to train the network.  
  
**• How is the loss being minimized?**  
The loss being minimized with zero gradient optimizer.  

In [ ]:
print(best_model)

SimpleCNN(
  (conv_layer1): Conv2d(3, 16, kernel_size=(5, 5), stride=(2, 2))
  (pool_layer1): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (conv_layer2): Conv2d(16, 16, kernel_size=(5, 5), stride=(2, 2))
  (pool_layer2): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (fully_connected_layer): Linear(in_features=1600, out_features=64, bias=True)
  (final_layer): Linear(in_features=64, out_features=11, bias=True)
)


- **What is the training accuracy for your network after training? What is the validation accuracy?**   
    
End of the default 200 epochs:         
Training ended returning the best model:  
-Best val acc: 0.4725274725274725,     
-Best val loss: 2.042543033758799
    
-Best train acc: 0.6444444444444445,   
-Best train loss: 1.913529396057129  


- **What do these two numbers tell you about what your network is doing?**  
Training accuracy shows that learning success of the best model when the model have whole dataset. Validation accuracy is a result of kind an exam after studying. Validation accuracy provides a learning process for new datasets.  


***1.3 Add Pooling Layers***     
*After you applied max pooling,* 
- What happened to your results?
- How did the training accuracy vs. validation accuracy change? What does that tell you about the effect of max pooling on your network?


In [ ]:
from CnnModel import SimpleCNN
from train import Trainer
from DataLoader import LoaderClass
import torch
from torch.nn import CrossEntropyLoss
from torch import optim
from torchvision import transforms, utils
import numpy as np
import os
from PIL import Image,ImageFile

LR = 1e-4
Momentum = 0.9 # If you use SGD with momentum
BATCH_SIZE = 16
USE_CUDA = True
POOLING = True  ### ADD POOLING LAYER
NUM_EPOCHS = 200
PATIENCE = -1
TRAIN_PERCENT = 0.8
VAL_PERCENT = 0.2
NUM_ARTISTS = 11
DATA_PATH = "./art_data/artists"
ImageFile.LOAD_TRUNCATED_IMAGES = True # Do not change this


def load_artist_data():
    data = []
    labels = []
    artists = [x for x in os.listdir(DATA_PATH) if x != '.DS_Store']
    print(artists)
    for folder in os.listdir(DATA_PATH):
        class_index = artists.index(folder)
        for image_name in os.listdir(DATA_PATH + "/" + folder):
            img = Image.open(DATA_PATH + "/" + folder + "/" + image_name)
            artist_label = (np.arange(NUM_ARTISTS) == class_index).astype(np.float32)
            data.append(np.array(img))
            labels.append(artist_label)
    shuffler = np.random.permutation(len(labels))
    data = np.array(data)[shuffler]
    labels = np.array(labels)[shuffler]

    length = len(data)
    val_size = int(length*0.2)
    val_data = data[0:val_size+1]
    train_data = data[val_size+1::]
    val_labels = labels[0:val_size+1]
    train_labels = labels[val_size+1::]
    print(val_labels)
    data_dict = {"train_data":train_data,"val_data":val_data}
    label_dict = {"train_labels":np.array(train_labels),"val_labels":np.array(val_labels)}

    return data_dict,label_dict

if __name__ == "__main__":
    data,labels = load_artist_data()
    model = SimpleCNN(use_cuda=True,pooling=POOLING)
    optimizer = optim.Adam(model.parameters(), lr=LR)
    transforms = {
        'train': transforms.Compose([
            transforms.Resize(50),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        ]),
        'val': transforms.Compose([
            transforms.Resize(50),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        ]),
        }
    train_dataset = LoaderClass(data,labels,"train",transforms["train"])
    valid_dataset = LoaderClass(data,labels,"val",transforms["val"])

    train_loader = torch.utils.data.DataLoader(train_dataset,
                                               batch_size=BATCH_SIZE,
                                               shuffle=True, num_workers=4, pin_memory=True)
    val_loader = torch.utils.data.DataLoader(valid_dataset,
                                             batch_size=BATCH_SIZE,
                                             shuffle=True, num_workers=4, pin_memory=True)

    criterion = CrossEntropyLoss()
    trainer_m = Trainer(model, criterion, train_loader, val_loader, optimizer, num_epoch=NUM_EPOCHS, patience=PATIENCE,batch_size=BATCH_SIZE,lr_scheduler= None)
    best_model_pool = trainer_m.train()

['richard wilson', 'paul cezanne', 'paul sandby', 'claude monet', 'peter paul rubens', 'canaletto', 'j. m. w. turner', 'paul gauguin', 'george romney', 'john robert cozens', 'rembrandt']
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 1. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


Epoch 1,phase: Training: 2it [00:00,  2.04it/s, val acc:  0.094,val loss:  2.396]  

New best loss, loss is:  2.396190603574117 acc is:  0.0989010989010989


Epoch 2,phase: Training: 3it [00:01,  1.46it/s, val acc:  0.094,val loss:  2.393]  

New best loss, loss is:  2.3928514321645102 acc is:  0.0989010989010989


Epoch 3,phase: Training: 4it [00:02,  1.27it/s, val acc:  0.167,val loss:  2.386]  

New best loss, loss is:  2.3857667446136475 acc is:  0.17582417582417584


Epoch 4,phase: Training: 5it [00:03,  1.18it/s, val acc:  0.219,val loss:  2.369]  

New best loss, loss is:  2.368627349535624 acc is:  0.23076923076923078


Epoch 5,phase: Training: 6it [00:04,  1.13it/s, val acc:  0.219,val loss:  2.331]  

New best loss, loss is:  2.3312778075536094 acc is:  0.23076923076923078


Epoch 6,phase: Training: 7it [00:05,  1.10it/s, val acc:  0.219,val loss:  2.271]  

New best loss, loss is:  2.270650625228882 acc is:  0.23076923076923078


Epoch 7,phase: Training: 8it [00:06,  1.08it/s, val acc:  0.219,val loss:  2.260]  

New best loss, loss is:  2.2597826719284058 acc is:  0.23076923076923078


Epoch 8,phase: Training: 9it [00:07,  1.06it/s, val acc:  0.260,val loss:  2.230]  

New best loss, loss is:  2.230235973993937 acc is:  0.27472527472527475


Epoch 9,phase: Training: 10it [00:08,  1.05it/s, val acc:  0.365,val loss:  2.205]  

New best loss, loss is:  2.2050723234812417 acc is:  0.38461538461538464


Epoch 10,phase: Training: 11it [00:09,  1.03it/s, val acc:  0.396,val loss:  2.160] 

New best loss, loss is:  2.1602930625279746 acc is:  0.4175824175824176


Epoch 11,phase: Training: 12it [00:10,  1.03it/s, val acc:  0.396,val loss:  2.144]  

New best loss, loss is:  2.143599510192871 acc is:  0.4175824175824176


Epoch 12,phase: Training: 13it [00:11,  1.03it/s, val acc:  0.406,val loss:  2.129]  

New best loss, loss is:  2.1290367046991983 acc is:  0.42857142857142855


Epoch 13,phase: Training: 14it [00:12,  1.03it/s, val acc:  0.417,val loss:  2.123]  

New best loss, loss is:  2.1227917671203613 acc is:  0.43956043956043955


Epoch 14,phase: Training: 15it [00:13,  1.03it/s, val acc:  0.417,val loss:  2.124]  

Not a better score


Epoch 15,phase: Training: 16it [00:14,  1.03it/s, val acc:  0.417,val loss:  2.113]  

New best loss, loss is:  2.112517555554708 acc is:  0.43956043956043955


Epoch 16,phase: Training: 17it [00:15,  1.03it/s, val acc:  0.427,val loss:  2.097]  

New best loss, loss is:  2.097330172856649 acc is:  0.45054945054945056


Epoch 17,phase: Training: 18it [00:16,  1.04it/s, val acc:  0.417,val loss:  2.109]  

Not a better score


Epoch 18,phase: Training: 19it [00:17,  1.04it/s, val acc:  0.417,val loss:  2.095]  

New best loss, loss is:  2.094654699166616 acc is:  0.43956043956043955


Epoch 19,phase: Training: 20it [00:18,  1.04it/s, val acc:  0.396,val loss:  2.099]  

Not a better score


Epoch 20,phase: Training: 21it [00:19,  1.03it/s, val acc:  0.406,val loss:  2.097]  

Not a better score


Epoch 21,phase: Training: 22it [00:20,  1.03it/s, val acc:  0.448,val loss:  2.086]  

New best loss, loss is:  2.0856314500172934 acc is:  0.4725274725274725


Epoch 22,phase: Training: 23it [00:21,  1.03it/s, val acc:  0.406,val loss:  2.095]  

Not a better score


Epoch 23,phase: Training: 24it [00:22,  1.03it/s, val acc:  0.427,val loss:  2.083]  

New best loss, loss is:  2.0831555922826133 acc is:  0.45054945054945056


Epoch 24,phase: Training: 25it [00:23,  1.04it/s, val acc:  0.458,val loss:  2.065]  

New best loss, loss is:  2.0650810400644937 acc is:  0.4835164835164835


Epoch 25,phase: Training: 26it [00:24,  1.03it/s, val acc:  0.458,val loss:  2.065]  

Not a better score


Epoch 26,phase: Training: 27it [00:25,  1.03it/s, val acc:  0.458,val loss:  2.073]  

Not a better score


Epoch 27,phase: Training: 28it [00:26,  1.04it/s, val acc:  0.448,val loss:  2.067]  

Not a better score


Epoch 28,phase: Training: 29it [00:27,  1.03it/s, val acc:  0.448,val loss:  2.058]  

New best loss, loss is:  2.0583136081695557 acc is:  0.4725274725274725


Epoch 29,phase: Training: 30it [00:28,  1.03it/s, val acc:  0.458,val loss:  2.050]  

New best loss, loss is:  2.0504230856895447 acc is:  0.4835164835164835


Epoch 30,phase: Training: 31it [00:29,  1.03it/s, val acc:  0.469,val loss:  2.059]  

Not a better score


Epoch 31,phase: Training: 32it [00:30,  1.11s/it, val acc:  0.448,val loss:  2.052]  

Not a better score


Epoch 32,phase: Training: 33it [00:32,  1.23s/it, val acc:  0.458,val loss:  2.056]  

Not a better score


Epoch 33,phase: Training: 34it [00:33,  1.31s/it, val acc:  0.490,val loss:  2.046]  

New best loss, loss is:  2.0462300181388855 acc is:  0.5164835164835165


Epoch 34,phase: Training: 35it [00:34,  1.21s/it, val acc:  0.469,val loss:  2.053]  

Not a better score


Epoch 35,phase: Training: 36it [00:35,  1.14s/it, val acc:  0.469,val loss:  2.043]  

New best loss, loss is:  2.043475091457367 acc is:  0.4945054945054945


Epoch 36,phase: Training: 37it [00:36,  1.11s/it, val acc:  0.469,val loss:  2.048]  

Not a better score


Epoch 37,phase: Training: 38it [00:37,  1.08s/it, val acc:  0.469,val loss:  2.038]  

New best loss, loss is:  2.037895898024241 acc is:  0.4945054945054945


Epoch 38,phase: Training: 39it [00:38,  1.05s/it, val acc:  0.469,val loss:  2.051]  

Not a better score


Epoch 38,phase: Validation: 39it [00:39,  1.05s/it, val acc:  0.500,val loss:  2.025]

In [ ]:
print(best_model_pool)

NameError: ignored

*1.4 Regularize the Network!*

Discuss your results. Which regularization technique was most effective?